dart/lib/src/sentry.dart

```diff
@@ -208,10 +208,15 @@ class Sentry {
     // try parsing the dsn
     Dsn.parse(options.dsn!);
 
-    // if logger os NoOp, let's set a logger that prints on the console
-    if (options.debug && options.logger == noOpLogger) {
-      options.logger = dartLogger;
+    // If the user set debug to false and the default logger is still
+    // `dartLogger` we set the logger to `noOpLogger`.
+    // `dartLogger` is the default because otherwise the logs won't be to the
+    // user visible between Sentry's initialization and the invocation of the
+    // options configuration callback.
+    if (options.debug == false && options.logger == dartLogger) {
+      options.logger = noOpLogger;
```

Comment:
Dart already has a `DiagnosticLogger`, when you set a `logger` it already wraps up to a `DiagnosticLogger` that takes into consideration the `diagnosticLevel` field.

Reply:
I think what Bruno meant is to rename the property to `diagnosticLogger` in order to make its purpose more clear.

Reply:
we cant rename now, this would require a major bump, every call to `options.logger(params)` would change.
what @bruno-garcia meant is, instead of doing such `if` condition to set a `dartLogger` or `noOpLogger ` could be encapsulated within a setter.

so options.debug = true (means setting `dartLogger`)
options.debug = false (means setting `noOpLogger`).

I like that.

the same could happen if one sets options.logger = function(), this would automatically set `options.debug = true` and options.logger = null would set `options.debug = false`, so such conditions are not across the SDK but only setting it enabled or not.

Reply:
> what @bruno-garcia meant is, instead of doing such if condition to set a dartLogger or noOpLogger could be encapsulated within a setter.

Yes! ;)

> the same could happen if one sets options.logger = function(), this would automatically set options.debug = true and options.logger = null would set options.debug = false

I wouldn't go this far though. One might change the logger type in an integration but not really be turning debug mode on or off.

> I think what Bruno meant is to rename the property to diagnosticLogger in order to make its purpose more clear.

Also this. What about we do this on a next major release? If so, is there a Discussion that keeps tracks of stuff we want to add to the next major?

> so options.debug = true (means setting dartLogger)
options.debug = false (means setting noOpLogger).

I'm not sure I would override the logger instance when someone sets `false` since next time you set it to `true` you no longer know what instance was set by the user before.

In .NET it's nullable, so that the call to the logger doesn't happen at all and hence no string concat or anything happens in order to invoke the logger. With NoOp you're calling things so any overhead to prepare the log entry is paid even though the logger doesn't do anything. That's just an optimization though, not saying it's something that we need to align. For really expensive operations we can always check: `logger.isEnabled(level)` before actually calling the logger.

Reply:
> If so, is there a Discussion that keeps tracks of stuff we want to add to the next major?

Now there is: https://github.com/getsentry/sentry-dart/discussions/419

Reply:
> With NoOp you're calling things so any overhead to prepare the log entry is paid even though the logger doesn't do anything. That's just an optimization though, not saying it's something that we need to align. For really expensive operations we can always check: logger.isEnabled(level) before actually calling the logger.

I feel like a callback for the logger would be more ergonomic than a null check.
For example:
```dart
logger.log(level, () => 'this callback gets invoked from the logger')
```
Than we would not need to have the actual code riddled with ifs and what not and also the logging is only computed when needen.

Reply:
There's an allocation for the callback though